File: examples/rnacontext/rnacontext_regr.R
hlp = "Post-processing of experiments with RBP binding affinity (RNAcontext dataset)."
require(optparse)
require(ggplot2)
require(scmamp)
require(xtable)

# Parse input arguments
option_list = list(
  make_option(c("-i", "--input"), type="character", help="Results file (.csv)"),
  make_option(c("-o", "--output"), type="character", help="Output directory.")
);
opt_parser = OptionParser(option_list=option_list, description=hlp);
opt = parse_args(opt_parser);
in_file = opt$input
out_dir = opt$output
dir.create(out_dir, showWarnings = FALSE)

# Read files
data = read.csv(in_file, stringsAsFactors = FALSE, header = TRUE)
data$dataset = unlist(lapply(strsplit(data$dataset, "_"), function(x) sprintf("%s (set %s)", x[1], x[4])))
data$dataset = gsub(".txt.gz", "", data$dataset)

# Cross-validation
data$id = paste(data$dataset, data$method, data$iteration, data$rank, sep=".")
agg = aggregate(data$evar_va, by=list(id=data$id), max)
row.names(agg) <- agg$id
data$best = agg[data$id, "x"] == data$evar_va
data = data[data$best,]
rank = unique(data$rank)

# Average
agg.m = aggregate(sqrt(data$mse), by=list(dataset=data$dataset, method=data$method), mean)
agg.s = aggregate(sqrt(data$mse), by=list(dataset=data$dataset, method=data$method), sd)

# 1 v 1 matrix
datasets = unique(data$dataset)
methods = unique(data$method)
R = matrix(NA, nrow=length(datasets), ncol=length(methods))
colnames(R) <- methods
row.names(R) <- datasets
for(i in 1:nrow(agg.m)) R[agg.m[i, "dataset"], agg.m[i, "method"]] = agg.m[i, "x"]

# Store CD plot
fname = file.path(out_dir, "CD_rank_evar.pdf")
pdf(fname)
plotCD(-R)
dev.off()

# Ranks
rnk = apply(R, 1, rank) 
message("Number of wins:")
rowSums(rnk == 1)

# Wilcoxon signed rank test
wilcox.test(R[,c("Mklaren")], R[,c("CSI")], paired = TRUE, alternative = "less")

# Store results in a table
M = matrix(Inf, ncol=length(datasets), nrow=length(methods))
S = matrix(Inf, ncol=length(datasets), nrow=length(methods))
row.names(M) = methods
colnames(M) = datasets
row.names(S) = methods
colnames(S) = datasets
for (i in 1:nrow(agg.m)) M[agg.m[i,"method"], agg.m[i,"dataset"]] = agg.m[i, "x"]
for (i in 1:nrow(agg.s)) S[agg.s[i,"method"], agg.s[i,"dataset"]] = agg.s[i, "x"]

# Text matrix 
Rt = matrix("", ncol=length(datasets), nrow=length(methods))
row.names(Rt) = methods
colnames(Rt) = datasets
for(j in 1:ncol(M)){
  Rt[names(M[,j]), colnames(M)[j]] = sprintf("%.3f$\\pm$%.3f", M[,j], S[,j])
  vals = round(M[methods, j], 3)
  best = names(which(vals == min(vals)))
  Rt[best, j] = sprintf("\\textbf{%.3f$\\pm$%.3f}", M[best,j], S[best,j])
}

# Append number of columns
dn = data[,c("dataset", "n")]
dn = dn[!duplicated(dn),]
row.names(dn) = dn$dataset
Rb = cbind(t(Rt), dn)
cols = c("n","Mklaren","CSI", "Nystrom", "ICD")
rows = order(Rb[,"n"])
Rb = Rb[,cols]
Rb = Rb[rows,]

# Store table
tab = xtable(Rb)
align(tab) = c("r", "r", "|", "l", "l", "l", "l")
fname = file.path(out_dir, "rmse.tex")
sink(fname)
print(tab,
      sanitize.colnames.function=identity,
      sanitize.text.function=identity)
sink()
message(sprintf("Written %s", fname))

File: examples/rnacontext/rnacontext_regr.py
hlp = """
    Experiments on RNAcontext string regression datasets. The methods Mklaren, CSI, Nystrom, ICD
    are run with string kernels to predict the RNA-binding protein binding affinity towards a given
    RNA sequence.
"""


import matplotlib
matplotlib.use("Agg")

import os
import csv
import time
import datetime
import argparse
import pickle, gzip
import scipy.stats as st
from mklaren.kernel.string_kernel import *
from mklaren.mkl.mklaren import Mklaren
from mklaren.kernel.kinterface import Kinterface
from mklaren.regression.ridge import RidgeLowRank
from datasets.rnacontext import load_rna, RNA_OPTIMAL_K, dataset2spectrum, RNA_DATASETS
from examples.inducing_points.inducing_points import meth2color
from examples.strings.string_utils import generic_function_plot


# Hyperparameters
rnk         = 5                                  # Approximation rank
lbd_range   = [0] + list(np.logspace(-5, 1, 7))  # Regularization parameter
iterations  = range(30)                          # Cross-validation iterations
delta       = 10                                 # Look-ahead parameter
n_tr        = 3000                               # Size of the training set of sequences
n_val       = 3000                               # Size of the validation set of sequences (fitting hyperparameters)


def process(dataset, outdir):
    """
    Run experiments with epcified parameters.
    :param dataset: Dataset key.
    :param outdir: Output directory.
    :return:
    """

    # List available kernels
    K_range = range(1, 11)
    kargs = [{"mode": SPECTRUM, "K": kl} for kl in K_range]
    kernels = ",".join(set(map(lambda t: t["mode"], kargs)))

    # Fixed settings
    methods = ["Mklaren", "CSI", "Nystrom", "ICD"]
    rank_range = (rnk,)
    trueK = RNA_OPTIMAL_K.get(dataset, None)

    # Fixed output
    # Create output directory
    detname = os.path.join(outdir, "_%s" % dataset)
    if not os.path.exists(outdir): os.makedirs(outdir)
    if not os.path.exists(detname): os.makedirs(detname)
    fname = os.path.join(outdir, "%s.csv" % dataset)
    print("Writing to %s ..." % fname)

    # Output
    header = ["dataset", "n", "L", "kernels", "method", "rank", "iteration", "lambda",
              "pivots", "time", "evar_tr", "evar_va", "evar", "mse"]
    fp = open(fname, "w", buffering=0)
    writer = csv.DictWriter(fp, fieldnames=header, quotechar='"', quoting=csv.QUOTE_ALL)
    writer.writeheader()

    # Load data
    data = load_rna(dataset)
    X = data["data"]
    y = st.zscore(data["target"])
    n, L = len(X), len(X[0])

    # Load feature spaces
    Ys = [pickle.load(gzip.open(dataset2spectrum(dataset, K))) for K in K_range]

    # Generate random datasets and perform prediction
    seed = 0
    for cv in iterations:

        # Select random test/train indices
        np.random.seed(seed)
        inxs = np.arange(n, dtype=int)
        np.random.shuffle(inxs)
        tr = inxs[:n_tr]
        va = inxs[n_tr:n_tr + n_val]
        te = inxs[n_tr + n_val:]

        # Training / test split
        y_tr = y[tr]
        y_va = y[va]
        y_te = y[te]

        # Print after dataset generation
        dat = datetime.datetime.now()
        print("%s\tdataset=%s cv=%d (computing kernels...)" % (dat, dataset, cv))

        # For plotting
        X_te = X[te]
        Ks = [Kinterface(kernel=string_kernel, data=X[tr], kernel_args=arg) for arg in kargs]

        # Precomputed kernel matrices
        Ls_tr = [np.array(Y[tr, :].dot(Y[tr, :].T).todense()) for Y in Ys]
        Ls_va = [np.array(Y[va, :].dot(Y[tr, :].T).todense()) for Y in Ys]
        Ls_te = [np.array(Y[te, :].dot(Y[tr, :].T).todense()) for Y in Ys]
        Ls_tr_sum = [sum(Ls_tr)]
        Ls_va_sum = [sum(Ls_va)]
        Ls_te_sum = [sum(Ls_te)]

        # Modeling
        for rank in rank_range:
            dat = datetime.datetime.now()
            print("\t%s\tdataset=%s cv=%d rank=%d" % (dat, dataset, cv, rank))
            best_models = {"True": {"y": y_te, "color": "black", "fmt": "--", }}
            for method in methods:
                best_models[method] = {"color": meth2color[method], "fmt": "-"}
                best_evar = -np.inf

                for lbd in lbd_range:
                    t1 = time.time()
                    if method == "Mklaren":
                        mkl = Mklaren(rank=rank, lbd=lbd, delta=delta)
                        try:
                            mkl.fit(Ls_tr, y_tr)
                            yt = mkl.predict(Xs=None, Ks=Ls_tr)
                            yv = mkl.predict(Xs=None, Ks=Ls_va)
                            yp = mkl.predict(Xs=None, Ks=Ls_te)
                            pivots = ",".join(map(lambda pi: str(K_range[pi]), mkl.G_mask.astype(int)))
                        except Exception as e:
                            print(e)
                            continue
                    else:
                        pivots = ""
                        if method == "CSI":
                            model = RidgeLowRank(rank=rank, method="csi",
                                                 method_init_args={"delta": delta}, lbd=lbd)
                        else:
                            model = RidgeLowRank(rank=rank, method=method.lower(), lbd=lbd)
                        try:
                            model.fit(Ls_tr_sum, y_tr)
                            yt = model.predict(Xs=None, Ks=Ls_tr_sum)
                            yv = model.predict(Xs=None, Ks=Ls_va_sum)
                            yp = model.predict(Xs=None, Ks=Ls_te_sum)
                        except Exception as e:
                            print(e)
                            continue
                    t2 = time.time() - t1

                    # Evaluate explained variance on the three sets
                    evar_tr = (np.var(y_tr) - np.var(yt - y_tr)) / np.var(y_tr)
                    evar_va = (np.var(y_va) - np.var(yv - y_va)) / np.var(y_va)
                    evar    = (np.var(y_te) - np.var(yp - y_te)) / np.var(y_te)
                    mse     = np.var(yp - y_te)

                    # Select best lambda to plot
                    if evar_va > best_evar:
                        best_evar = evar_va
                        best_yp = yp
                        best_models[method]["y"] = best_yp

                    # Write to output
                    row = {"L": L, "n": len(X), "method": method, "dataset": dataset,
                           "kernels": kernels, "rank": rank, "iteration": cv, "lambda": lbd,
                           "time": t2, "evar_tr": evar_tr, "evar_va": evar_va, "evar": evar,
                           "mse": mse, "pivots": pivots}

                    writer.writerow(row)
                    seed += 1

            # Plot a function fit after selecting best lambda
            fname = os.path.join(detname, "%s.generic_plot_cv-%d_rank-%d.pdf" % (dataset, cv, rank))
            generic_function_plot(f_out=fname, Ks=Ks, X=X_te,
                                  models=best_models,
                                  xlabel="K-mer length",
                                  xnames=K_range,
                                  truePar=K_range.index(trueK) if trueK else None)


if __name__ == "__main__":
    # Input arguments
    parser = argparse.ArgumentParser(description=hlp)
    parser.add_argument("dataset", help="Dataset. One of {%s}." % ", ".join(RNA_DATASETS))
    parser.add_argument("output",  help="Output directory.")
    args = parser.parse_args()

    # Output directory
    data_set = args.dataset
    out_dir = args.output
    assert data_set in RNA_DATASETS
    process(data_set, out_dir)